File: site/js/archetypes.js
/* "You play like Dwight Howard" - the NBA player a character most resembles.
 *
 * WHY A SHAPE AND NOT THE RATINGS THEMSELVES. A fourteen-year-old is rated in the twenties and
 * an All-Star in the nineties, so nearest-neighbour on raw numbers matches every character to
 * whoever is worst in the league - which is both useless and a bit insulting. Each player is
 * therefore centred on his OWN mean and scaled by his OWN spread: what he is good at relative to
 * himself. Cousins reads "scores inside, cannot shoot, rebounds" at any level, and so does a
 * prep kid built the same way.
 *
 * The comparison is cosine similarity on that shape, which ignores magnitude entirely - two
 * players with the same strengths in the same proportions score 1.0 however good either is.
 *
 * This mirrors tools/nba_archetypes.py exactly, field order included, because the JSON it writes
 * carries shapes computed there and a shape computed differently here would be compared against
 * them as though it were the same thing. The two implementations are checked against each other
 * in tests/test_archetypes.py rather than trusted to stay in step.
 *
 * Stamina and Fouling are deliberately absent: Stamina was set administratively to a flat 70 for
 * every character, so it carries no information about anybody, and Fouling is a habit rather than
 * an ability.
 */

/** The 16 rating fields the shape is built from, in the order the JSON's vectors use. */
export const SHAPE_FIELDS = [
  'InsideScoring', 'JumpShot', 'FtShot', '3pShot', '3pUsage', 'Handling', 'Passing',
  'PostDefense', 'PerimeterDefense', 'Stealing', 'Blocking', 'OReb', 'DReb',
  'Strength', 'Quickness', 'Jumping',
];

let CACHE = null;

/** The archetype file, fetched once. Null when it is missing - the page carries on without it. */
export function loadArchetypes(path = 'data/nba-archetypes.json') {
  if (!CACHE) {
    CACHE = fetch(path)
      .then((r) => (r.ok ? r.json() : null))
      .then((d) => (d && Array.isArray(d.players) ? d : null))
      .catch(() => null);
  }
  return CACHE;
}

/**
 * A ratings object as a shape vector, or null when he has no spread at all.
 *
 * Rounded to 4 decimals to match the python, so a value that lands exactly between two
 * candidates breaks the same way in both.
 */
export function shapeOf(ratings) {
  const values = SHAPE_FIELDS.map((k) => Number((ratings || {})[k]) || 0);
  const mean = values.reduce((a, b) => a + b, 0) / values.length;
  const spread = Math.sqrt(
    values.reduce((a, v) => a + (v - mean) ** 2, 0) / values.length);
  if (spread < 1e-6) return null;       // every rating identical: he has no shape to compare
  return values.map((v) => Math.round(((v - mean) / spread) * 10000) / 10000);
}

/** Cosine similarity, 1.0 being the same style of player. */
export function similarity(a, b) {
  let dot = 0;
  let na = 0;
  let nb = 0;
  for (let i = 0; i < a.length; i += 1) {
    dot += a[i] * b[i];
    na += a[i] * a[i];
    nb += b[i] * b[i];
  }
  return (na && nb) ? dot / (Math.sqrt(na) * Math.sqrt(nb)) : 0;
}

/**
 * The `count` NBA players a ratings sheet most resembles, best first.
 *
 * Ties are broken by name, so the same sheet always produces the same answer - two players can
 * sit at identical similarity to four decimal places, and "you play like" changing between two
 * reloads of the same character would read as the page being broken.
 */
export function matchArchetypes(ratings, data, count = 3) {
  if (!data || !Array.isArray(data.players)) return [];
  const mine = shapeOf(ratings);
  if (!mine) return [];
  return data.players
    .filter((p) => Array.isArray(p.shape) && p.shape.length === mine.length)
    .map((p) => ({ ...p, score: similarity(mine, p.shape) }))
    .sort((x, y) => y.score - x.score || x.name.localeCompare(y.name))
    .slice(0, count);
}

/** "scores inside, rebounds, blocks shots" - what the match is known for, from his top ratings. */
export function archetypeStrengths(player) {
  return (player.top || []).map(([field]) => field);
}

File: commissioner/archetypes.py
"""Which real NBA player does this character play like?

The comparison is honest because it is on the same scale: FBPB3 ships
`PlayerFiles/2015 Season Start.csv`, 579 real players rated in exactly the fields the codec
writes. `tools/nba_archetypes.py` turns that file into `site/data/nba-archetypes.json`; this
module answers questions against it, for the Discord card and anything else on this side. The
builder asks the same question in JavaScript against the same file.

STYLE, NOT LEVEL, which is the whole reason this is not a nearest-neighbour search. A
fourteen-year-old with everything in the twenties is not "most like" the worst player in the
league - but that is precisely what matching on raw ratings answers, for every character, so
they would all come back matched to the same few end-of-bench names and the feature would be
worthless. Each player is reduced to a SHAPE first: his ratings centred on his own mean and
scaled by his own spread, which says what he is good at RELATIVE TO HIMSELF. DeMarcus Cousins
reads as "scores inside, cannot shoot, rebounds" whether he is rated 98 or 45, and so does a
prep kid built the same way.
"""
from __future__ import annotations

import json
import math
from pathlib import Path

ROOT = Path(__file__).resolve().parents[1]
DATA = ROOT / "site" / "data" / "nba-archetypes.json"

# The ratings that describe HOW somebody plays. Stamina and Fouling are deliberately out: they
# say how long he lasts and how carelessly he defends, not what kind of player he is.
SHAPE = ["InsideScoring", "JumpShot", "FtShot", "3pShot", "3pUsage", "Handling", "Passing",
         "PostDefense", "PerimeterDefense", "Stealing", "Blocking", "OReb", "DReb",
         "Strength", "Quickness", "Jumping"]

_cache = {"at": None, "players": None}


def shape_of(ratings, fields=None):
    """Ratings as a shape. None when there is no spread at all - no real row, but a made-up
    one can be flat, and a zero vector would match everything equally."""
    keys = fields or SHAPE
    values = [float((ratings or {}).get(k) or 0) for k in keys]
    if not values:
        return None
    mean = sum(values) / len(values)
    spread = math.sqrt(sum((v - mean) ** 2 for v in values) / len(values))
    if spread < 1e-6:
        return None
    return [(v - mean) / spread for v in values]


def similarity(a, b):
    """Cosine similarity between two shapes; 1.0 is the same style of player."""
    if not a or not b or len(a) != len(b):
        return 0.0
    dot = sum(x * y for x, y in zip(a, b))
    na = math.sqrt(sum(x * x for x in a))
    nb = math.sqrt(sum(y * y for y in b))
    return dot / (na * nb) if na and nb else 0.0


def players(path=DATA):
    """The archetype list, read once and re-read only if the file changes underneath."""
    try:
        stamp = path.stat().st_mtime
    except OSError:
        return []
    if _cache["at"] != stamp:
        try:
            payload = json.loads(path.read_text(encoding="utf-8"))
        except (OSError, ValueError):
            return []
        _cache.update({"at": stamp, "players": payload.get("players") or [],
                       "fields": payload.get("fields") or SHAPE})
    return _cache["players"] or []


def most_like(ratings, count=3, path=DATA, pool=None):
    """The closest real players to a set of ratings, best first. [] when we cannot say.

    Never raises and never guesses: a missing archetype file, an empty one, or a character with
    no ratings all come back as "no answer", because a card that omits the line is fine and a
    card that names the wrong player is not.
    """
    # `pool` is for the tool that BUILDS the file: it has the players in memory, straight from
    # the roster CSV, and there may be no JSON on disk yet to read them back from.
    pool = pool if pool is not None else players(path)
    if not pool:
        return []
    mine = shape_of(ratings, _cache.get("fields"))
    if mine is None:
        return []
    scored = [(similarity(mine, p.get("shape")), p) for p in pool]
    scored.sort(key=lambda sp: -sp[0])
    return [{**p, "score": round(s, 3)} for s, p in scored[:count] if s > 0]
